operator/builtin/input/file/file_test.go

```diff
@@ -681,65 +686,182 @@ func TestMultiCopyTruncateSlow(t *testing.T) {
+			totalLines:      10,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Slow/NoDeletion",
+			totalLines:      20,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Slow/Deletion",
+			totalLines:      30,
+			maxLinesPerFile: 10,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
+		{
+			name:            "Slow/Deletion/ExceedFingerprint",
+			totalLines:      300,
+			maxLinesPerFile: 100,
+			maxBackupFiles:  1,
+			writeInterval:   3 * time.Millisecond,
+			pollInterval:    20 * time.Millisecond,
+		},
 	}
 
-	waitForMessages(t, logReceived, expected)
-}
-
-func TestRapidRotate_CopyTruncate(t *testing.T) {
-	t.Skip()
-	getMessage := func(m int) string { return fmt.Sprintf("message %d", m) }
-
-	numMessages := 1000
+	testRun := func(tc rotationTest, copyTruncate bool) func(t *testing.T) {
+		return func(t *testing.T) {
+			operator, logReceived, tempDir := newTestFileOperator(t,
+				func(cfg *InputConfig) {
+					cfg.PollInterval = helper.Duration{tc.pollInterval}
+				},
+				func(out *testutil.FakeOutput) {
+					out.Received = make(chan *entry.Entry, tc.totalLines)
+				},
+			)
+			logger := getRotatingLogger(t, tempDir, tc.maxLinesPerFile, tc.maxBackupFiles, copyTruncate)
+
+			expected := make([]string, 0, tc.totalLines)
+			for i := 0; i < tc.totalLines; i++ {
+				expected = append(expected, fmt.Sprintf("this is a fifty char log entry with the number %3d", i))
+			}
 
-	operator, logReceived, tempDir := newTestFileOperator(t, nil)
+			require.NoError(t, operator.Start())
+			defer operator.Stop()
 
-	logger := getRotatingLogger(t, tempDir, 10, true)
+			for _, message := range expected {
+				logger.Writer().Write([]byte(message + "\n"))
+				time.Sleep(tc.writeInterval)
+			}
 
-	expected := make([]string, 0, numMessages)
+			received := make([]string, 0, tc.totalLines)
+		LOOP:
+			for {
+				select {
+				case e := <-logReceived:
+					received = append(received, e.Record.(string))
+				case <-time.After(250 * time.Millisecond):
+					break LOOP
+				}
+			}
 
-	for i := 0; i < numMessages; i++ {
-		expected = append(expected, getMessage(i))
+			if tc.ephemeralLines {
+				if !tc.expectEphemeralLines() {
+					// This is helpful for test development, and ensures the sample computation is used
+					t.Logf("Potentially unstable ephemerality expectation for test: %s", tc.name)
+				}
+				require.Subset(t, expected, received)
+			} else {
+				require.ElementsMatch(t, expected, received)
+			}
+		}
 	}
 
-	require.NoError(t, operator.Start())
-	defer operator.Stop()
-
-	for _, message := range expected {
-		logger.Writer().Write([]byte(message + "\n"))
-		time.Sleep(time.Millisecond)
+	for _, tc := range cases {
```

Comment:
How much time do these tests add to our test runtime?

Reply:
It's definitely not trivial, but total runtime the package (for me locally) is ~10 seconds now.

I tuned it in quite a bit, but ultimately we need to write a decent amount of lines with reasonable timing in between in order to accommodate execution in different environments. I'm open to suggestions on what else we can do to reduce this, as long as we're confident that we're testing the right cases and doing so in at least as stable a manner.

Reply:
I'm happy with 10 seconds for now. These tests seem to add a lot of value, so it's worth the increased testing time. It's not too hard to scope down tests with `go test` anyways